crates/stoffel-vm/src/net/program_sync.rs

```diff
@@ -0,0 +1,173 @@
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub enum ProgramSyncMessage {
+    ProgramAnnounce {
+        party_id: PartyId,
+        program_id: [u8; 32],
+        size: u64,
+        entry: String,
+    },
+    ProgramAck {
+        party_id: PartyId,
+        program_id: [u8; 32],
+    },
+    ProgramFetchRequest {
+        program_id: [u8; 32],
+    },
+    ProgramBytes {
+        program_id: [u8; 32],
+        bytes: Vec<u8>,
+    },
+    ProgramComplete {
+        program_id: [u8; 32],
+    },
+}
+
+/// Returns the cache directory for storing synced programs
+pub fn cache_dir() -> PathBuf {
+    std::env::var("STOFFEL_CACHE")
+        .map(PathBuf::from)
+        .unwrap_or_else(|_| dirs::home_dir().unwrap_or_else(|| ".".into()).join(".stoffel").join("programs"))
+}
+
+/// Returns the path where a program with the given ID should be cached
+pub fn program_path(program_id: &[u8; 32]) -> PathBuf {
+    let hex = hex::encode(program_id);
+    cache_dir().join(hex)
+}
+
+/// Computes a BLAKE3 hash of the program bytes to use as its ID
+pub fn program_id_from_bytes(bytes: &[u8]) -> [u8; 32] {
+    let mut hasher = Hasher::new();
+    hasher.update(b"stoffel-program-v1");
+    hasher.update(bytes);
+    hasher.finalize().as_bytes().clone()
+}
+
+/// Ensures the cache directory exists
+pub fn ensure_cache_dir() -> Result<(), String> {
+    let dir = cache_dir();
+    fs::create_dir_all(&dir).map_err(|e| e.to_string())
+}
+
+/// Sends a control message to a peer using stoffelnet's simple send/receive
+pub async fn send_ctrl(conn: &dyn PeerConnection, msg: &ProgramSyncMessage) -> Result<(), String> {
+    let bytes = bincode::serialize(msg).map_err(|e| e.to_string())?;
+    conn.send(&bytes).await
+}
+
+/// Receives a control message from a peer
+pub async fn recv_ctrl(conn: &dyn PeerConnection) -> Result<ProgramSyncMessage, String> {
+    let buf = conn.receive().await?;
+    let msg: ProgramSyncMessage = bincode::deserialize(&buf).map_err(|e| e.to_string())?;
+    Ok(msg)
+}
+
+/// Sends program bytecode to a peer
+pub async fn send_program_bytes(
+    conn: &dyn PeerConnection,
+    program_id: [u8; 32],
+    bytes: Arc<Vec<u8>>,
+) -> Result<(), String> {
+    let msg = ProgramSyncMessage::ProgramBytes { program_id, bytes: bytes.to_vec() };
+    send_ctrl(conn, &msg).await
+}
+
+/// Receives program bytecode from a peer
+pub async fn recv_program_bytes(conn: &dyn PeerConnection, expected_id: [u8; 32]) -> Result<Vec<u8>, String> {
+    let msg = recv_ctrl(conn).await?;
+    match msg {
+        ProgramSyncMessage::ProgramBytes { program_id, bytes } => {
+            if program_id != expected_id {
+                return Err("program_id mismatch".into());
+            }
+            Ok(bytes)
+        }
+        _ => Err("expected ProgramBytes message".into()),
+    }
+}
+
+/// High-level helper to ensure all parties agree on the program and those who don't have it fetch it.
+pub async fn agree_and_sync_program(
```

Comment:
In some circumstances, they will need to fetch the program from elsewhere like a smart contract or from a url or something else